### PR TITLE
fix: allow experiments before authentication [ROAD-472]

### PR DESCRIPTION
**Pull request by bastiandoetsch**
Previously, variants were only fetched from Amplitude when a user was already authenticate. This is not enough, when we want to experiment before authentication.

Now the variants are fetched on startup even if there is no token, using the anonymousID from segment as deviceId in the Amplitude Experiment API as additional parameter. Thus, when userID is empty, the deviceId is still filled and experiments can be assigned to users.

### CODE REVIEW

**Comment by bastiandoetsch** [comment]
will fix this on monday

File: src/main/kotlin/snyk/amplitude/AmplitudeExperimentService.kt
Comment (michelkaporin) [inline src/main/kotlin/snyk/amplitude/AmplitudeExperimentService.kt:22]: Why change it to `treatment` if we had it defined as `test` for the previous experiment? I'd rather rename the group on Amplitude side.

Also there's possibility that we will be having more groups when running A/B/C test in future.

File: src/main/kotlin/snyk/amplitude/AmplitudeExperimentService.kt
Comment (michelkaporin) [inline src/main/kotlin/snyk/amplitude/AmplitudeExperimentService.kt:23]: Out of curiosity - what's the semantic difference between having storage outside and inside companion object? 👀

File: src/main/kotlin/snyk/amplitude/api/AmplitudeExperimentApiClient.kt
Comment (michelkaporin) [inline src/main/kotlin/snyk/amplitude/api/AmplitudeExperimentApiClient.kt:23]: Why would we even try to fetch the variants if both properties are blank? This will lead to undefined behaviour and wouldn't count towards the experiment data.

File: src/main/resources/application.properties
Comment (michelkaporin) [inline src/main/resources/application.properties:1]: ⛔

**Comment by michelkaporin** [comment]
@bastiandoetsch let's link up the ticket here

File: src/main/kotlin/snyk/amplitude/AmplitudeExperimentService.kt
Comment (bastiandoetsch) [reply src/main/kotlin/snyk/amplitude/AmplitudeExperimentService.kt:23]: Static storage vs local storage

File: src/main/kotlin/snyk/amplitude/api/AmplitudeExperimentApiClient.kt
Comment (bastiandoetsch) [reply src/main/kotlin/snyk/amplitude/api/AmplitudeExperimentApiClient.kt:23]: Because we still fetch it with a deviceID, so we don't need the userID.

File: src/main/resources/application.properties
Comment (bastiandoetsch) [reply src/main/resources/application.properties:1]: Argh.

File: src/main/kotlin/snyk/amplitude/AmplitudeExperimentService.kt
Comment (michelkaporin) [reply src/main/kotlin/snyk/amplitude/AmplitudeExperimentService.kt:23]: thank you!

File: src/main/kotlin/snyk/amplitude/api/AmplitudeExperimentApiClient.kt
Comment (michelkaporin) [reply src/main/kotlin/snyk/amplitude/api/AmplitudeExperimentApiClient.kt:23]: let me ask you differently - what's the purpose of fetching it with a blank device id and blank userId?

File: src/main/kotlin/snyk/amplitude/AmplitudeExperimentService.kt
Comment (bastiandoetsch) [reply src/main/kotlin/snyk/amplitude/AmplitudeExperimentService.kt:22]: Treatment is the default on Amplitude - and I'd rather stick to their default. But in the end, I don't care much.

File: src/main/kotlin/snyk/amplitude/api/AmplitudeExperimentApiClient.kt
Comment (bastiandoetsch) [reply src/main/kotlin/snyk/amplitude/api/AmplitudeExperimentApiClient.kt:23]: The deviceID is never blank, as it is automatically populated on instance creation with the default value.

File: build.gradle.kts
Comment (ArtsiomCh) [inline build.gradle.kts:35]: What's the reason to update versions above? 
Are we sure we don't get new bugs with such update?

File: src/main/kotlin/io/snyk/plugin/SnykPostStartupActivity.kt
Comment (ArtsiomCh) [inline src/main/kotlin/io/snyk/plugin/SnykPostStartupActivity.kt:77]: minor: could be `val publicUserId = if ...`

**Review by ArtsiomCh — APPROVED** [review]
LGTM
Please address minor notes from my side.
Also please get approval from @michelkaporin before merging as he know better how all that Amplitude's experiment magic works.

File: build.gradle.kts
Comment (bastiandoetsch) [reply build.gradle.kts:35]: Yeah, it's only a minor upgrade and I tested it also manually against amplitude and verified events are sent.

**Review by michelkaporin — APPROVED** [review]

**Pull request merged**
